### ISSUE TITLE
Allow defining `fmap` for different input types with pattern matching

### ISSUE THREAD

**Issue by fakuivan** [evhub/coconut#653]
Right now the dunder method `__fmap__` is used for classes to define functor-like behaviour. It'd be nice if it would also be possible to augment library classes with an implementation of `fmap` that lives outside the module the class was defined.

My specific use case is for sympy classes, specially the `Equality` class. I currently define `map_sides` a function that takes a function and an equality and applies that function to both sides, so the equality is preserved.

```coconut
def map_sides(f, eq: sp.Eq): -> sp.Eq
    return sp.Eq(f(eq.lhs), f(eq.rhs))
```

If `fmap` were to be defined as a pattern-matching function, the following would be possible.

```coconut
addpattern def fmap(func, eq is sp.Eq) -> sp.Eq:
    return sp.Eq(f(eq.lhs), f(eq.rhs))
```

**Comment by evhub** [comment]
Currently, on `coconut-develop` (`pip install -U coconut-develop`), you can do
```python
Coconut Interpreter v2.0.0-a_dev48:
(enter 'exit()' or press Ctrl-D to end)
>>> @addpattern$(?, fmap)
    match def fmap(f, int(x)) = f(x)

>>> fmap((.+1), 5)
6
>>> fmap((.+1), (1,2,3))
(2, 3, 4)
>>>
```

Does that solve your use case?

**Comment by evhub** [comment]
@fakuivan Not supporting dotted names there is a bug—should be fixed now in `coconut-develop>=2.0.0-a_dev49`.

Coconut not recognizing argument ordering in `sp.Eq` is because it doesn't define a [PEP 634](https://peps.python.org/pep-0634/#class-patterns) compliant `__match_args__`. You can either bug `sympy` about adding `__match_args__` to their classes, use `sp.Eq(lhs=..., rhs=...)` syntax, or patch in a `__match_args__` yourself with something like `sp.Eq.__match_args__ = ("lhs", "rhs")`.

**Comment by fakuivan** [comment]
Great! I'll close this since the functionality should be added in 2.0. Many thanks!
